x/stake/keeper/validator.go

```diff
@@ -191,132 +191,148 @@ func (k Keeper) ClearTendermintUpdates(ctx sdk.Context) {
 
 //___________________________________________________________________________
 
-// perfom all the nessisary steps for when a validator changes its power
-// updates all validator stores as well as tendermint update store
-// may kick out validators if new validator is entering the bonded validator group
+// Perfom all the nessisary steps for when a validator changes its power.  This
+// function updates all validator stores as well as tendermint update store.
+// It may kick out validators if new validator is entering the bonded validator
+// group.
+//
 // nolint: gocyclo
 // TODO: Remove above nolint, function needs to be simplified
 func (k Keeper) UpdateValidator(ctx sdk.Context, validator types.Validator) types.Validator {
 	store := ctx.KVStore(k.storeKey)
 	pool := k.GetPool(ctx)
-	ownerAddr := validator.Owner
+	oldValidator, oldFound := k.GetValidator(ctx, validator.Owner)
 
-	// always update the main list ordered by owner address before exiting
-	defer func() {
-		k.SetValidator(ctx, validator)
-	}()
+	validator = k.updateForRevoking(ctx, oldFound, oldValidator, validator)
+	powerIncreasing := k.getPowerIncreasing(ctx, oldFound, oldValidator, validator)
+	validator.BondHeight, validator.BondIntraTxCounter = k.bondIncrement(ctx, oldFound, oldValidator, validator)
+	valPower := k.updateValidatorPower(ctx, oldFound, oldValidator, validator, pool)
+	cliffPower := k.GetCliffValidatorPower(ctx)
 
-	// retrieve the old validator record
-	oldValidator, oldFound := k.GetValidator(ctx, ownerAddr)
+	switch {
+	// if already bonded and power increasing only need to update tendermint
+	case powerIncreasing && !validator.Revoked &&
+		(oldFound && oldValidator.Status() == sdk.Bonded):
+
+		bz := k.cdc.MustMarshalBinary(validator.ABCIValidator())
+		store.Set(GetTendermintUpdatesKey(validator.Owner), bz)
+
+	// if is a new validator and the new power is less than the cliff validator
+	case cliffPower != nil && !oldFound &&
+		bytes.Compare(valPower, cliffPower) == -1: //(valPower < cliffPower
+		// skip to completion
+
+	// if was unbonded and the new power is less than the cliff validator
+	case cliffPower != nil &&
+		(oldFound && oldValidator.Status() == sdk.Unbonded) &&
+		bytes.Compare(valPower, cliffPower) == -1: //(valPower < cliffPower
+		// skip to completion
+
+	default:
```

Comment:
Add comment with the known state (validator was unbonded/not found & has power greater than the cliff validator, or was bonded and has decreased in power)